LabApiUtilities/src/main/com/microsoft/identity/labapi/utilities/authentication/KeyVaultAuthenticationClient.java

```diff
@@ -0,0 +1,109 @@
+    private final static String SCOPE = "https://vault.azure.net/.default";
+    private final static String CERTIFICATE_ALIAS = "AutomationRunner";
+    private final static String KEYSTORE_TYPE = "Windows-MY";
+    private final static String KEYSTORE_PROVIDER = "SunMSCAPI";
+
+    // tenant id where lab api and key vault api is registered
+    private final static String TENANT_ID = "72f988bf-86f1-41af-91ab-2d7cd011db47";
+
+    private final static String AUTHORITY = "https://login.microsoftonline.com/" + TENANT_ID;
+
+    private final IConfidentialAuthClient mConfidentialAuthClient;
+    private final String mClientSecret;
+
+    public KeyVaultAuthenticationClient(@NonNull final IConfidentialAuthClient confidentialAuthClient) {
+        mConfidentialAuthClient = confidentialAuthClient;
+        mClientSecret = null;
+    }
+
+    public KeyVaultAuthenticationClient(@NonNull final IConfidentialAuthClient confidentialAuthClient,
+                                        @NonNull final String clientSecret) {
+        mConfidentialAuthClient = confidentialAuthClient;
+        mClientSecret = clientSecret;
+    }
+
+    private ITokenParameters getTokenParametersForKeyVault() {
+        return TokenParameters
+                .builder()
+                .clientId(CLIENT_ID)
+                .authority(AUTHORITY)
+                .scope(SCOPE)
+                .build();
+    }
+
+    @Override
+    public String getAccessToken() throws LabApiException {
+        final IAuthenticationResult authenticationResult;
+
+        if (mClientSecret != null && mClientSecret.trim().length() > 0) {
+            // if client secret is provided then we would use that to acquire token
+            authenticationResult = mConfidentialAuthClient.acquireToken(
+                    mClientSecret, getTokenParametersForKeyVault()
+            );
+        } else {
+            // client secret was not provided...so we would try to use the Certificate
+            // the cert must be in store for it to succeed..otherwise it would just fail
+            final KeyStoreConfiguration keyStoreConfiguration = new KeyStoreConfiguration(
+                    KEYSTORE_TYPE, KEYSTORE_PROVIDER, null
+            );
+
+            final ClientCertificateMetadata certificateMetadata = new ClientCertificateMetadata(
+                    CERTIFICATE_ALIAS, null
+            );
+
+            final CertificateCredential certificateCredential = CertificateCredential.create(
+                    keyStoreConfiguration,
+                    certificateMetadata
+            );
+
+            authenticationResult = mConfidentialAuthClient.acquireToken(
+                    certificateCredential, getTokenParametersForKeyVault()
+            );
+        }
+
+        if (authenticationResult != null) {
+            return authenticationResult.getAccessToken();
+        } else {
+            throw new LabApiException(LabError.FAILED_TO_GET_TOKEN_FOR_KEYVAULT);
```

Comment:
When this exception is thrown, if you're trying to figure out what happened, it would be nice if it threw from one branch or the other in the above if statement, or contained some indication of which one it failed in.

Reply:
Address comment: https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/1389/commits/d5f60ffe283865adb9f375ba910ffcc948f8d775